### PR TITLE
[FLINK-14656] [table-planner-blink] blink planner should also fetch catalog statistics for permanent CatalogTableImpl

### PR DESCRIPTION
**Pull request by godfreyhe**
## What is the purpose of the change

*let blink planner also fetch catalog statistics for permanent CatalogTableImpl*

## Brief change log

  - *handle stats for CatalogTableImpl in DatabaseCalciteSchema#convertPermanentTable method*

## Verifying this change

  - *Extended CatalogStatisticsTest for permanent CatalogTableImpl*

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / **not documented**)

### CODE REVIEW

File: flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/catalog/DatabaseCalciteSchema.java [hunk 1]
```diff
@@ -136,7 +136,7 @@ private Table convertTemporaryTable(
 				return convertSinkTable(connectorTable);
 			}
 		} else if (table instanceof CatalogTable) {
-			return convertCatalogTable(tablePath, (CatalogTable) table, null);
+			return convertCatalogTable(tablePath, (CatalogTable) table, null, TableStats.UNKNOWN);
```
Comment (KurtYoung) [inline flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/catalog/DatabaseCalciteSchema.java]: Is is possible for temporary tables also have table stats? 
@dawidwys What do you think about this?

Comment (JingsongLi) [reply flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/catalog/DatabaseCalciteSchema.java]: Here if we take a look to `extractTableStats`. It gets stats from catalog. I think `ConnectorCatalogTable` should never be put into catalog.
So I think we need remove `extractTableStats` of `ConnectorCatalogTable`.

Comment (KurtYoung) [reply flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/catalog/DatabaseCalciteSchema.java]: Yes, this is current implementation. But from user's perspective, when they registered a temporary table, it should also have an opportunity to have table stats.

Comment (JingsongLi) [reply flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/catalog/DatabaseCalciteSchema.java]: But the stats of temporary table should not come from catalog, right?
Another thing is that now temporary table is `CatalogTable` too.
`ConnectorCatalogTable` only come from deprecated `registerTableSource`.

Comment (godfreyhe) [reply flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/catalog/DatabaseCalciteSchema.java]: That is also my concern: users could also store  a `ConnectorCatalogTable` instance into a `Catalog` by `Catalog#createTable` (not by `TableEnvironment#fromTableSource` or `TableEnvironment#registerTableSource`) if the `ConnectorCatalogTable` instance is serializable (`GenericInMemoryCatalog` even does not need serializable object). should we prevent users from storing `ConnectorCatalogTable` instance into a `Catalog` ? or should we clear that `ConnectorCatalogTable` could only be used for temporary table ?

Comment (godfreyhe) [reply flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/catalog/DatabaseCalciteSchema.java]: In old planner, `ConnectorCatalogTable` may be a permanent table(from catalog) too in `DatabaseCalciteSchema`.

**Comment by JingsongLi** [comment]
Hi @godfreyhe @KurtYoung , I would like to suggest that we should generate the `TableStats` directly before `convertPermanentTable`. Why distinguish the tables? View should also have `TableStats`, right?

**Comment by godfreyhe** [comment]
@JingsongLi , I think `View` should not have `TableStats`

File: flink-table/flink-table-planner-blink/src/test/java/org/apache/flink/table/planner/catalog/CatalogStatisticsTest.java [hunk 2]
```diff
@@ -59,32 +63,75 @@
...
+		catalog.createTable(
+				ObjectPath.fromString("default_database.T2"),
+				new CatalogTableImpl(tableSchema, properties, ""),
+				false);
+
+		alterTableStatistics(catalog);
+
+		Table table = tEnv.sqlQuery("select * from T1, T2 where T1.s3 = T2.s3");
+		String result = tEnv.explain(table);
+		// T1 is broadcast side
```
Comment (KurtYoung) [inline flink-table/flink-table-planner-blink/src/test/java/org/apache/flink/table/planner/catalog/CatalogStatisticsTest.java]: Do we have other choices for validating the table stats? Relying on small table will be broadcasted in a join is kind of fragile IMO.

Comment (godfreyhe) [reply flink-table/flink-table-planner-blink/src/test/java/org/apache/flink/table/planner/catalog/CatalogStatisticsTest.java]: i think one choice is using metadata handlers to verify the table stats.

Comment (KurtYoung) [reply flink-table/flink-table-planner-blink/src/test/java/org/apache/flink/table/planner/catalog/CatalogStatisticsTest.java]: Yes, let's test this via metadata handlers.

**Pull request merged**